source/postprocess/visualization/boundary_velocity_residual.cc

```diff
@@ -0,0 +1,243 @@
+      {
+        Assert ((computed_quantities[0].size() == dim), ExcInternalError());
+        auto cell = input_data.template get_cell<DoFHandler<dim> >();
+
+        const double velocity_scaling_factor =
+          this->convert_output_to_years() ? year_in_seconds : 1.0;
+
+        for (unsigned int q=0; q<computed_quantities.size(); ++q)
+          for (unsigned int d = 0; d < dim; ++d)
+            computed_quantities[q](d)= 0;
+
+        // We only want the output at the top boundary, so only compute it if the current cell
+        // has a face at the top boundary.
+        bool cell_at_top_boundary = false;
+        for (unsigned int f=0; f<GeometryInfo<dim>::faces_per_cell; ++f)
+          if (cell->at_boundary(f) &&
+              (this->get_geometry_model().translate_id_to_symbol_name (cell->face(f)->boundary_id()) == "top"))
+            cell_at_top_boundary = true;
+
+
+        if (cell_at_top_boundary)
+          {
+            for (unsigned int q=0; q<computed_quantities.size(); ++q)
+              {
+                Tensor<1,dim> data_velocity;
+                if (use_spherical_unit_vectors && use_ascii_data)
+                  {
+                    const std::array<double,dim> cartesian_position = this->get_geometry_model().
+                                                                      cartesian_to_natural_coordinates(input_data.evaluation_points[q]);
+
+                    Point<dim> internal_position;
+
+                    for (unsigned int d = 0; d < dim; ++d)
+                      internal_position[d] = cartesian_position[d];
+
+                    for (unsigned int d = 0; d < dim; ++d)
+                      data_velocity[d] = ascii_data_lookup->get_data(internal_position, d);
+
+                    data_velocity = Utilities::Coordinates::spherical_to_cartesian_vector(data_velocity, input_data.evaluation_points[q]);
+
+                    for (unsigned int d = 0; d < dim; ++d)
+                      computed_quantities[q](d) = (data_velocity[d] - input_data.solution_values[q][d] * velocity_scaling_factor) ;
+                  }
+                else if (use_ascii_data && !use_spherical_unit_vectors)
+                  {
+                    for (unsigned int d = 0; d < dim; ++d)
+                      {
+                        data_velocity[d] = ascii_data_lookup->get_data(input_data.evaluation_points[q], d);
+                        computed_quantities[q](d) = (data_velocity[d] - input_data.solution_values[q][d] * velocity_scaling_factor) ;
+                      }
+                  }
+                else
+                  {
+                    data_velocity =  gplates_lookup->surface_velocity(input_data.evaluation_points[q]);
+                    for (unsigned int d = 0; d < dim; ++d)
+                      computed_quantities[q](d) = (data_velocity[d] - input_data.solution_values[q][d]) * velocity_scaling_factor;
+                  }
```

Comment:
I would write this approximately the following way. This way you do not need 3 branches, but only ascii data or gplates. Also it seems that at the moment you only support either : (i) sperical models and spherical unit vectors for velocity (branch 1), or (ii) cartesian models and no spherical unit vectors (branch 2). But with my version you also support ascii data files for spherical models, but prescribing cartesian velocities in the data columns.

```suggestion
if (use_ascii_data == true)
{
  Point<dim> internal_position = input_data.evaluation_points[q];
  
  if (this->get_geometry_model().natural_coordinate_system() == Utilities::Coordinates::spherical)
  {
    const std::array<double,dim> spherical_position = this->get_geometry_model().
      cartesian_to_natural_coordinates(input_data.evaluation_points[q]);

  for (unsigned int d = 0; d < dim; ++d)
    internal_position[d] = spherical_position[d];
  }

  for (unsigned int d = 0; d < dim; ++d)
    data_velocity[d] = ascii_data_lookup->get_data(internal_position, d);

  if (use_spherical_unit_vectors == true)
    data_velocity = Utilities::Coordinates::spherical_to_cartesian_vector(data_velocity, input_data.evaluation_points[q]);
}
else
{
  data_velocity =  gplates_lookup->surface_velocity(input_data.evaluation_points[q]);
}
                  
for (unsigned int d = 0; d < dim; ++d)
  computed_quantities[q](d) = (data_velocity[d] - input_data.solution_values[q][d]) * velocity_scaling_factor;
```